File: py/pv_ephemeral_ids/allocator.py
# py/pv_ephemeral_ids/allocator.py

# PORTED BY CHATGPT, NOT TESTED

import random
from typing import Dict, List, Set

from .generator import generate_id_map


class EphemeralIds:
    def __init__(self, id_map: Dict[str, List[str]]):
        self.id_map = id_map
        self.reset()

        if self.num_available == 0:
            raise ValueError("No available identifiers.")

        # All starters should be the same length
        self.starter_length = len(next(iter(self.starters)))

    def reset(self):
        # Full list of available starter tokens
        self.starters: List[str] = list(self.id_map.keys())
        self.num_available = len(self.starters)
        self.active_ids: Set[str] = set()

    def create(self) -> str:
        """Allocates a new ephemeral identifier."""
        if self.num_available == 0:
            raise RuntimeError("No available identifiers.")

        # Randomly pick a starter
        index = random.randint(0, self.num_available - 1)
        starter = self.starters[index]
        suffixes = self.id_map[starter]
        suffix = random.choice(suffixes)
        eid = f"{starter}{suffix}"

        # Swap out the starter to mark it as used
        self.starters[index] = self.starters[self.num_available - 1]
        self.num_available -= 1

        self.active_ids.add(eid)
        return eid

    def release(self, eid: str):
        """Releases a previously allocated identifier."""
        if eid not in self.active_ids:
            return  # silent skip
        self.active_ids.remove(eid)

        starter = eid[:self.starter_length]
        if self.num_available < len(self.starters):
            self.starters[self.num_available] = starter
        else:
            self.starters.append(starter)
        self.num_available += 1

    @classmethod
    async def from_repo(cls, model_repo: str, prefix: str = "", long: bool = False, cache: bool = True):
        id_map = generate_id_map(model_repo, prefix=prefix, long=long, cache=cache)
        return cls(id_map)

# end of file

File: js/src/allocator.js
/* js/src/allocator.js */

import { generateIdMap } from './generator.js';

export class EphemeralIds {
  /**
   * Create a new EphemeralIds instance.
   * @param {Object<string, string[]>} idMap - Map of identifier starters to available suffixes.
   */
  constructor(idMap) {
    this.idMap = idMap;
    this.reset();

    if (this.numAvailable === 0) {
        throw new Error('No available identifiers.');
    }

    // Length of the starter tokens
    this.starterLength = this.starters[0].length;
  }

  reset() {
    // Create an array of all the available starter tokens
    this.starters = [...Object.keys(this.idMap)];
    this.numAvailable = this.starters.length;

    // Set of active identifiers
    this.activeIds = new Set();
  }

  /**
   * Allocates a new ephemeral identifier. This reserves the starter token to this identifier.
   * @returns {string}
   */
  create() {
    // If there are no available identifiers, throw an error
    if (this.numAvailable === 0) {
      throw new Error('No available identifiers.');
    }

    // Random number between 0 and numAvailable - 1
    const randomIndex = Math.floor(Math.random() * this.numAvailable);
    const starter = this.starters[randomIndex];
    const suffixes = this.idMap[starter];
    const suffixIndex = Math.floor(Math.random() * suffixes.length);
    const suffix = suffixes[suffixIndex];
    const id = `${starter}${suffix}`;
    
    // Swap the selected starter to the end of the available entries,
    // and reduce the number of available identifiers
    this.starters[randomIndex] = this.starters[this.numAvailable - 1];
    // this.starters[this.numAvailable - 1] = starter; // Just discard
    this.numAvailable--;

    // Add the identifier to the active set
    this.activeIds.add(id);

    return id;
  }

  /**
   * Releases a previously allocated identifier.
   * @param {string} id
   */
  release(id) {
    // Check if it's actually an active id, and remove it from the set
    if (!this.activeIds.has(id)) {
      return; // Silent skip
    }
    this.activeIds.delete(id);

    // Extract the starter token from the identifier
    // and add it back to the available starters
    const starter = id.slice(0, this.starterLength);
    this.starters[this.numAvailable] = starter;
    this.numAvailable++;
  }

  /**
   * Async factory method to create an EphemeralIds instance from a model repo.
   * @param {string} modelRepo - Hugging Face model repo
   * @param {Object} [options]
   * @param {string} [options.prefix]
   * @param {boolean} [options.long]
   * @param {boolean} [options.cache]
   * @returns {Promise<EphemeralIds>}
   */
  static async fromRepo(modelRepo, options = {}) {
    const idMap = await generateIdMap(modelRepo, options);
    return new EphemeralIds(idMap);
  }
}

/* end of file */
